### PR TITLE
Initial migration to v6

### PR DESCRIPTION
**Pull request by WilliamBZA**

### CODE REVIEW

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/EnrichPreV44MessagesWithHostDetailsMutator.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            }
+
+            return Task.FromResult(0);
+        }
+
+        Guid HostId;
+        string Host;
+    }
+
+    class EnrichPreV44MessagesWithHostDetailsMutatorInitializer : INeedInitialization
```
Comment (WilliamBZA) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/EnrichPreV44MessagesWithHostDetailsMutator.cs]: Not sure about this. Couldn't constructor inject in the INeedInitialization instance, but needed the context to get the ReadOnlySettings, so split it out.

Is there a better way?

File: packaging/nuget/servicecontrol.plugin.Nsb6.heartbeat.nuspec [hunk 2]
```diff
@@ -15,10 +15,10 @@
     <copyright>$copyright$</copyright>
     <tags>nservicebus servicebus msmq cqrs publish subscribe</tags>
     <dependencies>
-		    <dependency id="NServiceBus" version="[5.0.0.0, 6.0.0.0)" />
+		    <dependency id="NServiceBus" version="[6.0.0-unstable, 7.0.0)" />
     </dependencies>
   </metadata>
   <files>
-    <file src="..\..\src\ServiceControl.Plugin.Nsb5.Heartbeat\bin\Release\ServiceControl.Plugin.Nsb5.Heartbeat.*" target="lib\net45" />
+    <file src="..\..\src\ServiceControl.Plugin.Nsb6.Heartbeat\bin\Release\ServiceControl.Plugin.Nsb6.Heartbeat.*" target="lib\net45" />
```
Comment (danielmarbach) [inline packaging/nuget/servicecontrol.plugin.Nsb6.heartbeat.nuspec]: should it be `lib\net452`?

File: src/ServiceControl.Plugin.Nsb6.Heartbeat.AcceptanceTests/app.config [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat.AcceptanceTests/app.config]: I think we can ditch this

File: src/ServiceControl.Plugin.Nsb6.Heartbeat.Sample/App.config [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
+    <section name="AuditConfig" type="NServiceBus.Config.AuditConfig, NServiceBus.Core"/>
+    <section name="UnicastBusConfig" type="NServiceBus.Config.UnicastBusConfig, NServiceBus.Core"/>
   </configSections>
-  <MessageForwardingInCaseOfFaultConfig ErrorQueue="error" />
-  <AuditConfig QueueName="audit" />
+  <MessageForwardingInCaseOfFaultConfig ErrorQueue="error"/>
+  <AuditConfig QueueName="audit"/>
   <startup>
-    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5" />
+    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5.2"/>
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat.Sample/App.config]: I think we can ditch this startup section

File: src/ServiceControl.Plugin.Nsb6.Heartbeat.Sample/App.config [hunk 5]
```diff
@@ -19,9 +19,9 @@
   <runtime>
     <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
       <dependentAssembly>
-        <assemblyIdentity name="NServiceBus.Core" publicKeyToken="9fc386479f8a226c" culture="neutral" />
-        <bindingRedirect oldVersion="0.0.0.0-5.0.0.0" newVersion="5.0.0.0" />
+        <assemblyIdentity name="NServiceBus.Core" publicKeyToken="9fc386479f8a226c" culture="neutral"/>
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat.Sample/App.config]: Do we need this? If yes it should be 6.0.0

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs [hunk 6]
```diff
@@ -82,33 +85,33 @@ protected override void OnStop()
...
-                base.OnStop();
+                return Task.FromResult(0);
             }
 
-            void NotifyEndpointStartup(HostInformation hostInfo, DateTime startupTime)
+            void NotifyEndpointStartup(DateTime startupTime)
             {
                 // don't block here since StartupTasks are executed synchronously.
-                Task.Run(() => SendEndpointStartupMessage(hostInfo, startupTime, cancellationTokenSource.Token));
+                Task.Run(() => SendEndpointStartupMessage(startupTime, cancellationTokenSource.Token));
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: We don't need this `Task.Run` just don't await the Task returned from `SendEndpointStartupMessage` and use `TaskEx.Ignore()`

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs [hunk 7]
```diff
@@ -82,33 +85,33 @@ protected override void OnStop()
...
-                            Host = hostInfo.DisplayName,
+                            HostId = HostId,
+                            Host = Host,
                             Endpoint = endpointName,
-                            HostDisplayName = hostInfo.DisplayName,
-                            HostProperties = hostInfo.Properties,
+                            HostDisplayName = Host,
+                            HostProperties = Properties,
                             StartedAt = startupTime
                         }, ttlTimeSpan);
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: `ConfigureAwait(false)`

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs [hunk 8]
```diff
@@ -116,24 +119,24 @@ void SendEndpointStartupMessage(HostInformation hostInfo, DateTime startupTime,
...
 
-                    Task.Delay(registrationRetryInterval, cancellationToken)
-                        .ContinueWith(t => SendEndpointStartupMessage(hostInfo, startupTime, cancellationToken), cancellationToken);
+                    await Task.Delay(registrationRetryInterval, cancellationToken).ConfigureAwait(false);
+                    await SendEndpointStartupMessage(startupTime, cancellationToken).ConfigureAwait(false);
                 }
             }
 
-            void SendHeartbeatMessage(HostInformation hostInfo)
+            async void SendHeartbeatMessage()
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: Yieks! Return a task. This is 100 [Enter exercise here which hurts]. :dancers:

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs [hunk 9]
```diff
@@ -116,24 +119,24 @@ void SendEndpointStartupMessage(HostInformation hostInfo, DateTime startupTime,
...
-                    Host = hostInfo.DisplayName,
-                    HostId = hostInfo.HostId
+                    Host = Host,
+                    HostId = HostId
                 };
 
                 try
                 {
-                    backend.Send(heartBeat, ttlTimeSpan);
+                    await backend.Send(heartBeat, ttlTimeSpan);
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: `ConfigureAwait(false)`

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs [hunk 10]
```diff
@@ -82,33 +85,33 @@ protected override void OnStop()
...
-                Task.Run(() => SendEndpointStartupMessage(hostInfo, startupTime, cancellationTokenSource.Token));
+                Task.Run(() => SendEndpointStartupMessage(startupTime, cancellationTokenSource.Token));
             }
 
-            void StartHeartbeats(HostInformation hostInfo)
+            void StartHeartbeats()
             {
                 Logger.DebugFormat("Start sending heartbeats every {0}", heartbeatInterval);
-                heartbeatTimer = new Timer(x => SendHeartbeatMessage(hostInfo), null, TimeSpan.Zero, heartbeatInterval);
+                heartbeatTimer = new Timer(x => SendHeartbeatMessage(), null, TimeSpan.Zero, heartbeatInterval);
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: You probably need an async timer here as well. Check out the custom checks. Ping me if you need help

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Internal/RepeatedFailuresOverTimeCircuitBreaker.cs [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            this.name = name;
+            this.triggerAction = triggerAction;
+            this.timeToWaitBeforeTriggering = timeToWaitBeforeTriggering;
+
+            timer = new Timer(CircuitBreakerTriggered);
+        }
+
+        public void Dispose()
+        {
+            //Injected
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Internal/RepeatedFailuresOverTimeCircuitBreaker.cs]: This only works if you have `Fody.Janitor` installed

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs (hunk 8 above)
Comment (WilliamBZA) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: Shhhhhhhh

https://media.giphy.com/media/XItRQJP0wai7m/giphy.gif

Comment (WilliamBZA) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/Heartbeats.cs]: Shhhhhhh

you didnt see anything

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+        string GetServiceControlAddress()
+        {
+            var queueName = ConfigurationManager.AppSettings[@"ServiceControl/Queue"];
+            if (!string.IsNullOrEmpty(queueName))
+            {
+                return queueName;
+            }
+
+            // TODO: Is this right??
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs]: can you remove?

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/TaskExtendionMethods.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Threading.Tasks;
+
+namespace ServiceControl.Plugin.Nsb6.Heartbeat
+{
+    /// <summary>
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/TaskExtendionMethods.cs]: Probably not necessary

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/TaskExtendionMethods.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    /// Extension methods here because different namespace makes the NServiceBus.TaskEx not visible
+    /// </summary>
+    static class TaskExtensionMethods
+    {
+        /// <summary>
+        /// Ignore a task, let it run without waiting
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/TaskExtendionMethods.cs]: Probably not necessary

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/VersionChecker.cs [hunk 15]
```diff
@@ -14,7 +14,7 @@ static VersionChecker()
                 fileVersion.FileBuildPart);
         }
 
-        public static Version CoreFileVersion { get; set; }
+        private static Version CoreFileVersion { get; }
```
Comment (danielmarbach) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/VersionChecker.cs]: Redundant private?

**Comment by mikeminutillo** [comment]
@WilliamBZA @SzymonPobiega @danielmarbach is this one ready for review?

**Comment by danielmarbach** [comment]
As far as I see from a functionality perspective this is done. We reviewed it as well. We cannot merge because this repo follows Githubflow

**Comment by johnsimons** [comment]
"make Heartbeats feature public" why ?

**Comment by johnsimons** [comment]
The code that figures our where to send heartbeats to does not take into consideration the new api to configure audit and error:

```
var faultsForwarderConfig = settings.GetConfigSection<MessageForwardingInCaseOfFaultConfig>();
```

this code expects a config section but how about if user uses the new code only Api ?

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+            }
+
+            string errorAddress;
+            if (TryGetErrorQueueAddress(out errorAddress))
+            {
+                var qm = Parse(errorAddress);
+                return "Particular.ServiceControl" + "@" + qm.Item2;
+            }
+
+            if (VersionChecker.CoreVersionIsAtLeast(4, 1))
```
Comment (johnsimons) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs]: I don't think we need this any more ?

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+            if (!string.IsNullOrEmpty(queueName))
+            {
+                return queueName;
+            }
+
+            string errorAddress;
+            if (TryGetErrorQueueAddress(out errorAddress))
+            {
+                var qm = Parse(errorAddress);
+                return "Particular.ServiceControl" + "@" + qm.Item2;
```
Comment (johnsimons) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs:115]: The `@` symbol, is this safe to assume that all transports are going to work ?

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+            }
+
+            if (VersionChecker.CoreVersionIsAtLeast(4, 1))
+            {
+                //audit config was added in 4.1
+                string address;
+                if (TryGetAuditAddress(out address))
+                {
+                    var qm = Parse(errorAddress);
+                    return "Particular.ServiceControl" + "@" + qm.Item2;
```
Comment (johnsimons) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs]: The `@` symbol, is this safe to assume that all transports are going to work ?

File: src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        static Tuple<string, string> Parse(string destination)
+        {
+            if (string.IsNullOrEmpty(destination))
+            {
+                throw new ArgumentException("Invalid destination address specified", nameof(destination));
+            }
+
+            var arr = destination.Split('@');
```
Comment (johnsimons) [inline src/ServiceControl.Plugin.Nsb6.Heartbeat/ServiceControlBackend.cs:149]: The `@` symbol, is this safe to assume that all transports are going to work ?

**Pull request merged**